### PR TITLE
Merge type stubs for torch.nn.parallel

### LINKED ISSUE #91648: Merge type stubs for torch.nn.parallel
### 🐛 Describe the bug

[The `DataParallel` type stubs as they exist now](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/data_parallel.pyi#L6-L14) are pretty bare bones. They do not include stubs for:

- `forward`
- `replicate`
- `scatter`
- `parallel_apply`
- `gather`

https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/data_parallel.py#L150-L184

---

This leaves derived classes of `DataParallel` looking to extend these methods via `super()` with type checking errors.

```python
from typing import Iterable, List

from torch import nn


class MyDataParallel(nn.DataParallel):
    def replicate(self, module: nn.Module, device_ids: Iterable[int]) -> List[nn.Module]:
        replicas = super().replicate(module, device_ids)
        ...
        return replicas

```

```
[misc] mypy(error): "replicate" undefined in superclass
Pylance[reportGeneralTypeIssues]: Object of type "Tensor" is not callable
```

### Versions

```console
$ python collect_env.py 
Collecting environment information...
PyTorch version: 1.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (x86_64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.25.1
Libc version: N/A

Python version: 3.10.6 (main, Oct 24 2022, 11:04:34) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.991
[pip3] mypy-boto3-s3==1.26.0.post1
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] torch==1.13.0
[pip3] torchdata==0.5.0
[pip3] torchmetrics==0.11.0
[pip3] torchvision==0.14.0
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] torch                     1.13.0                   pypi_0    pypi
[conda] torchdata                 0.5.0                    pypi_0    pypi
[conda] torchmetrics              0.11.0                   pypi_0    pypi
[conda] torchvision               0.14.0                   pypi_0    pypi
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by ringohoffman**
Fixes #91648

As explained in the tracking issue, the incomplete type stubs in `torch/nn/parallel` mask `DataParallel` methods relevant for subclassing and also mask type issues present in the code as well.

There are various renamings and `type: ignore` comments added to bypass errors that arose from the merging.

@ezyang

### CODE REVIEW

File: torch/nn/parallel/scatter_gather.py [hunk 1]
```diff
@@ -43,11 +59,17 @@ def scatter_map(obj):
     try:
         res = scatter_map(inputs)
     finally:
-        scatter_map = None
+        scatter_map = None  # type: ignore[assignment]
     return res
 
 
-def scatter_kwargs(inputs, kwargs, target_gpus, dim=0):
+# TODO More precise types here.
```
Comment (Skylion007) [inline torch/nn/parallel/scatter_gather.py:66]: FYI: https://github.com/Instagram/MonkeyType can help you find more precise types here (if you remove the Any annotations).

**Comment by ezyang** [comment]
added albanD for public API errors

**Review by ezyang — APPROVED** [review]
Seems fine, but you'll have to figure out how to shut up the public API errors. Probably a missing `__all__` somewhere.

**Comment by ringohoffman** [comment]
Looks like the BC linter didn't like changing `get_a_var` to be private to appease the public interface linter (https://github.com/pytorch/pytorch/pull/101528/commits/ceeb242bc7a8291c2f108659e46fd81d418380ec). LMK if we can suppress this or if you want to include `get_a_var` in `__all__` in `torch/nn/parallel/parallel_apply.py` (it is only referenced in that file).

**Comment by ringohoffman** [comment]
@ezyang anything outstanding here?

**Comment by ringohoffman** [comment]
> It's still failing lintrunner, gotta fix the rest of the type errors in other files. You can do it locally with `lintrunner --all-files` since the reduced file heuristic doesn't work quite right for mypy.

@ezyang I addressed these errors here: https://github.com/pytorch/pytorch/pull/102046

**Comment by kit1980** [comment]
The bc_linter actually flagged this on the PR:

GitHub Actions
/ bc_linter
Function get_a_var: function deleted

**Comment by ringohoffman** [comment]
@kit1980 #102194
